instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/StateMapping.java

```diff
@@ -0,0 +1,150 @@
+  private final Set<String> stateKeys;
+
+  private StateMapping(
+      @Nullable String defaultState, Map<String, String> stateMapping, Set<String> stateKeys) {
+    this.defaultStateKey = defaultState;
+    this.stateMapping = stateMapping;
+    this.stateKeys = stateKeys;
+  }
+
+  /**
+   * Returns {@literal} true when empty, {@literal false} otherwise
+   *
+   * @return {@literal true} when state mapping is empty, {@literal false} otherwise
+   */
+  public boolean isEmpty() {
+    return stateKeys.isEmpty();
+  }
+
+  /**
+   * get state keys
+   *
+   * @return set of state keys, including the default one
+   */
+  public Set<String> getStateKeys() {
+    return stateKeys;
+  }
+
+  /**
+   * get default state key
+   *
+   * @return default state key, {@literal null} when empty
+   */
+  @Nullable
+  public String getDefaultStateKey() {
+    return defaultStateKey;
+  }
+
+  /**
+   * Get mapped state value
+   *
+   * @param rawValue raw state value from JMX attribute
+   * @return mapped state value from raw value, default value or {@literal null} when empty
+   */
+  @Nullable
+  public String getStateValue(String rawValue) {
+    String value = stateMapping.get(rawValue);
+    if (value == null) {
+      value = defaultStateKey;
+    }
+    return value;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Empty instance
+   *
+   * @return an empty {@link StateMapping} instance
+   */
+  public static StateMapping empty() {
+    return EMPTY;
+  }
+
+  public static class Builder {
+
+    private String defaultState;
+    private final Map<String, String> valueMapping;
+    private final Set<String> stateKeys;
+
+    private Builder() {
+      this.valueMapping = new HashMap<>();
+      this.stateKeys = new HashSet<>();
+    }
+
+    /**
+     * Adds default state key
+     *
+     * @param state state key
+     * @return this
+     */
+    @CanIgnoreReturnValue
+    public Builder withDefaultState(String state) {
+      if (defaultState != null) {
+        throw new IllegalStateException("default state already set");
+      }
+      defaultState = state;
+      stateKeys.add(defaultState);
+      return this;
+    }
+
+    /**
+     * Adds a mapped state value
+     *
+     * @param value raw value to be mapped
+     * @param state state value to map raw value to
+     * @return this
+     */
+    @CanIgnoreReturnValue
+    public Builder withMappedValue(String value, String state) {
+      String currentMapping = valueMapping.putIfAbsent(value, state);
+      if (currentMapping != null) {
+        throw new IllegalStateException(value + " already mapped to " + currentMapping);
+      }
+      stateKeys.add(state);
+      return this;
+    }
+
+    StateMapping build() {
+      if (stateKeys.isEmpty()) {
+        return EMPTY;
+      }
+
+      if (defaultState == null) {
+        throw new IllegalStateException("missing default state");
```

Comment:
Is requiring the default state necessary? I understand that if it is missing and the actual state string value is not in the list, the metric value will be zero for all combinations of provided metric attributes.

Reply:
I agree that this might be "best practice" rather than strict requirement, but this would likely be the case for all the mappings embedded in the agent anyway, so preventing invalid metrics and making mappings more future proof is probably a better option, I'm fine to remove this as a strict requirement though.